### PR TITLE
Support integer non-decimal notations

### LINKED ISSUE #599: Support binary/octal/hexadecimal numbers
- PHP Depend version: 2.10.3
- PHP Version: 8.1.0
- Installation type: composer

### Parsed code
```php
class Foo
{
    const BAR = 0xf;
}
```
### Current Behavior
When finding `ASTConstantDeclarator` and inspecting the value, we get:
`int(0)`

### Expected Behavior
`int(15)`

### Cause
On `Tokens::T_LNUMBER`, the parse set the value using `$defaultValue->setValue($signed * (int) $token->image);` but here `$token->image` is `0xf` and `(int) '0xf'` results int `0`.

### Resolution
- Parse `bBoOxX` integer to convert those structures to proper binary/octal/hexadecimal.
- The same way `_` should be ignored.
- `oO` should be disallowed before PHP 8.1 (nice to have).
- `_` should be disallowed before PHP 7.4 (nice to have).

### PR DESCRIPTION
**Pull request by kylekatarnls**
Type: feature
Issue: #556 Fix #599
Breaking change: no

### CODE REVIEW

File: src/test/php/PDepend/Source/Language/PHP/PHPParserVersion54Test.php [hunk 1]
```diff
@@ -73,20 +73,6 @@ public function testParserHandlesBinaryIntegerLiteral()
...
-    /**
-     * testParserThrowsExceptionForInvalidBinaryIntegerLiteral
-     *
-     * @return void
-     * @expectedException \PDepend\Source\Parser\UnexpectedTokenException
-     */
-    public function testParserThrowsExceptionForInvalidBinaryIntegerLiteral()
-    {
-        if (version_compare(phpversion(), '5.4alpha') >= 0) {
-            $this->markTestSkipped('This test only affects PHP < 5.4');
```
Comment (ravage84) [inline src/test/php/PDepend/Source/Language/PHP/PHPParserVersion54Test.php:85]: Good riddance!

File: src/main/php/PDepend/Source/Language/PHP/AbstractPHPParser.php [hunk 2]
```diff
@@ -244,6 +245,21 @@ abstract class AbstractPHPParser
...
+                       0
+                       |
+                       [1-9][0-9]*
+                       |
+                       0[xX][0-9a-fA-F]+
+                       |
+                       0[0-7]+
+                       |
+                       0[bB][01]+
+                     )$/x';
```
Comment (ravage84) [inline src/main/php/PDepend/Source/Language/PHP/AbstractPHPParser.php:261]: Looks like modern art, this way 😃

File: src/main/php/PDepend/Source/Language/PHP/AbstractPHPParser.php [hunk 3]
```diff
@@ -7880,4 +7902,37 @@ protected function parseThrowExpression()
     {
         throw $this->getUnexpectedTokenException();
     }
+
+    /**
+     * @param string $numberRepresentation
```
Comment (ravage84) [inline src/main/php/PDepend/Source/Language/PHP/AbstractPHPParser.php:7907]: A short description with an example of what goes in and what gets parsed (output) would be helpful to understan it.

File: src/main/php/PDepend/Source/Language/PHP/PHPParserVersion54.php [hunk 4]
```diff
@@ -242,8 +242,13 @@ protected function parseArray(ASTArray $array, $static = false)
     protected function parseIntegerNumber()
     {
         $token = $this->consumeToken(Tokens::T_LNUMBER);
+        $number = $token->image;
```
Comment (ravage84) [inline src/main/php/PDepend/Source/Language/PHP/PHPParserVersion54.php:245]: This improves readability 👍

File: src/test/php/PDepend/Source/AST/ASTLiteralTest.php [hunk 5]
```diff
@@ -217,22 +217,6 @@ public function testLiteralWithNonZeroBinaryIntegerValue()
...
-     *
-     * @return void
-     * @since 1.0.0
-     * @covers \PDepend\Source\Language\PHP\PHPParserVersion54
-     * @expectedException \PDepend\Source\Parser\UnexpectedTokenException
-     */
-    public function testLiteralWithBrokenBinaryIntegerThrowsExpectedException()
-    {
-        if (version_compare(phpversion(), '5.4alpha') >= 0) {
-            $this->markTestSkipped('This test only affects PHP < 5.4');
```
Comment (ravage84) [inline src/test/php/PDepend/Source/AST/ASTLiteralTest.php:231]: Good riddance!

File: src/main/php/PDepend/Source/Language/PHP/AbstractPHPParser.php (hunk 2 above)
Comment (kylekatarnls) [reply src/main/php/PDepend/Source/Language/PHP/AbstractPHPParser.php:261]: For the record, it's strictly conform to PHP doc:
https://www.php.net/manual/en/language.types.integer.php

Slightly different in PHP < 8.1 to remove the `[oO]?` for octal notation and in PHP < 7.4 to remove `_` support.

File: src/main/php/PDepend/Source/Language/PHP/AbstractPHPParser.php (hunk 3 above)
Comment (kylekatarnls) [reply src/main/php/PDepend/Source/Language/PHP/AbstractPHPParser.php:7907]:
```suggestion
     * @param string $numberRepresentation integer number as it appears in the code, `0xfe4`, `1_000_000`
```

File: src/main/php/PDepend/Source/Language/PHP/AbstractPHPParser.php (hunk 2 above)
Comment (tvbeek) [reply src/main/php/PDepend/Source/Language/PHP/AbstractPHPParser.php:261]: Maybe it is an idea to add a link to that documentation in the comment? (For the person reading it next year and thinking how did you write this)

Comment (kylekatarnls) [reply src/main/php/PDepend/Source/Language/PHP/AbstractPHPParser.php:261]:
```suggestion
    /**
     * Integer representation as allowed before PHP 7.4.
     * @see https://github.com/php/doc-en/blob/528e97348f45b473ca8ecea9997be1e268703562/language/types/integer.xml#L74-L86
     */
    const REGEXP_INTEGER = '/^(
                       0
                       |
                       [1-9][0-9]*
                       |
                       0[xX][0-9a-fA-F]+
                       |
                       0[0-7]+
                       |
                       0[bB][01]+
                     )$/x';
```

File: src/main/php/PDepend/Source/Language/PHP/PHPParserVersion74.php [hunk 6]
```diff
@@ -60,6 +60,21 @@
  */
 abstract class PHPParserVersion74 extends PHPParserVersion73
 {
+    /**
+     * Regular expression for integer numbers representation.
```
Comment (kylekatarnls) [inline src/main/php/PDepend/Source/Language/PHP/PHPParserVersion74.php:64]:
```suggestion
     * Regular expression for integer numbers representation.
     * (Add support for octal explicit notation.)
     *
     * @see https://php.net/manual/en/language.types.integer.php
     * @see https://github.com/php/doc-en/blob/085c38d45e466691062b4444c71f4dbe4198f884/language/types/integer.xml#L79-L91
```

File: src/main/php/PDepend/Source/Language/PHP/PHPParserVersion81.php [hunk 7]
```diff
@@ -58,6 +58,21 @@
  */
 abstract class PHPParserVersion81 extends PHPParserVersion80
 {
+    /**
+     * Regular expression for integer numbers representation.
```
Comment (kylekatarnls) [inline src/main/php/PDepend/Source/Language/PHP/PHPParserVersion81.php:62]:
```suggestion
     * Regular expression for integer numbers representation.
     * (PHP 7.4 added support for underscores, octal explicit notation still disallowed.)
     *
     * @see https://github.com/php/doc-en/blob/d494ffa4d9f83b60fe66972ec2c0cf0301513b4a/language/types/integer.xml#L77-L89
```

File: src/main/php/PDepend/Source/Language/PHP/AbstractPHPParser.php [hunk 8]
```diff
@@ -244,6 +245,21 @@ abstract class AbstractPHPParser
                        (array)\(\s*\)\s+
                      )ix';
 
+    /**
+     * Regular expression for integer numbers representation.
```
Comment (kylekatarnls) [inline src/main/php/PDepend/Source/Language/PHP/AbstractPHPParser.php:249]:
```suggestion
     * Regular expression for integer numbers representation.
     * (Underscore and explicit octal notation were not allowed before PHP 7.4.)
     *
     * @see https://github.com/php/doc-en/blob/528e97348f45b473ca8ecea9997be1e268703562/language/types/integer.xml#L74-L86
```

**Review by tvbeek — APPROVED** [review]

**Pull request merged**